match/lib/match/encryption/encryption.rb

```diff
@@ -0,0 +1,148 @@
+    # IV is randomly generated and provided unencrypted
+    # salt should be randomly generated and provided unencrypted (like in the current implementation)
+    # key is generated with OpenSSL::KDF::pbkdf2_hmac with properly chosen parameters
+    # Short explanation about salt and IV: https://stackoverflow.com/a/1950674/6324550
+    class EncryptionV2
+      ALGORITHM = 'aes-256-gcm'
+
+      def encrypt(data, password, salt)
+        cipher = ::OpenSSL::Cipher.new(ALGORITHM)
+        cipher.encrypt
+
+        keyivgen(cipher, password, salt)
+
+        encrypted_data = cipher.update(data)
+        encrypted_data << cipher.final
+
+        auth_tag = cipher.auth_tag
+
+        { encrypted_data: encrypted_data, auth_tag: auth_tag }
+      end
+
+      def decrypt(encrypted_data, password, salt, auth_tag)
+        cipher = ::OpenSSL::Cipher.new(ALGORITHM)
+        cipher.decrypt
+
+        keyivgen(cipher, password, salt)
+
+        cipher.auth_tag = auth_tag
+
+        data = cipher.update(encrypted_data)
+        data << cipher.final
+      end
+
+      private
+
+      def keyivgen(cipher, password, salt)
+        keyIv = ::OpenSSL::KDF.pbkdf2_hmac(password, salt: salt, iterations: 10_000, length: 32 + 12 + 24, hash: "sha256")
+        key = keyIv[0..31]
+        iv = keyIv[32..43]
+        auth_data = keyIv[44..-1]
+        cipher.key = key
+        cipher.iv = iv
+        cipher.auth_data = auth_data
+      end
+    end
+
+    class MatchDataEncryption
+      V1_PREFIX = "Salted__"
+      V2_PREFIX = "2_Salted__"
+
+      def encrypt(data, password, version = 2)
+        salt = SecureRandom.random_bytes(8)
+        if version == 2
+          e = EncryptionV2.new
+          encryption = e.encrypt(data, password, salt)
+          encrypted_data = V2_PREFIX + salt + encryption[:auth_tag] + encryption[:encrypted_data]
+        else
+          e = EncryptionV1.new
+          encrypted_data = V1_PREFIX + salt + e.encrypt(data, password, salt)
+        end
+        Base64.encode64(encrypted_data)
+      end
+
+      # expected_salt is only used when testing
+      def decrypt(base64encoded_encrypted, password)
+        stored_data = Base64.decode64(base64encoded_encrypted)
+        if stored_data.start_with?(V2_PREFIX)
+          salt = stored_data[10..17]
+          auth_tag = stored_data[18..33]
+          data_to_decrypt = stored_data[34..-1]
+          e = EncryptionV2.new
+          e.decrypt(data_to_decrypt, password, salt, auth_tag)
+        else
+          salt = stored_data[8..15]
+          data_to_decrypt = stored_data[16..-1]
+          e = EncryptionV1.new
+          begin
+            e.decrypt(data_to_decrypt, password, salt)
+          rescue => _ex
+            # Note that we are not guaranteed to catch the decryption errors here if the password is wrong
+            # as there's no integrity checks.
+            # With a wrong password, there's a 0.4% chance it will decrypt garbage and not fail
```

Comment:
Would be good to have the link to that discussion on the OpenSSL issue tracker here too in case anyone wonders about this in the future